Add executable bit configuration to allow `in_repo` and `on_disk` permissions to diverge

isuffix opened the pull request:
Let's try this again!

This is a remaking of #4478 that builds off of the many smaller PRs I've been making the past month.

One key difference from the approach in #4478 is how the third commit starts out by preserving the `in_repo` executable bit on all platforms, which makes the later commit adding `on_disk` for Unix easier to reason about.

Closes #3949

It took me a while to come to the final representation of the executable bit, but I think it's correct for what we want.

I find the symmetry between the `from_repo` and `from_disk` methods particularly convincing: if we're respecting the executable bit, then `on_disk` and `in_repo` always update to match each other. If we're ignoring the executable bit, then the two values are allowed to diverge, and instead we set them to their previous values (if known).

# Checklist

If applicable:

- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [x] I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes

isuffix commented on lib/src/local_working_copy.rs:
```diff
@@ -1771,22 +1787,13 @@ impl TreeState {
-    }
-
-    #[cfg_attr(windows, allow(unused_variables))]
-    fn set_executable(&self, disk_path: &Path, executable: bool) -> Result<(), CheckoutError> {
```
How about I just make `set_executable()` a top-level function that takes `ExecBit`, but placed just below the `ExecBit` impl?

yuja replied:
Yeah, that seems good. I don't remember why the current `set_executable()` takes `&self`, but there might be a plan to make it noop depending on configuration.

isuffix commented on lib/src/local_working_copy.rs:
```diff
@@ -127,31 +128,22 @@ use crate::working_copy::WorkingCopyFactory;
+        #[cfg(windows)]
+        let executable = prev_exec.map_or(false, |prev| prev.0);
+        Self(executable)
     }
```
We start to share logic between platforms in the next commit, so I'd rather leave it as-is.
(I'm marking resolved, feel free to unresolve if you want to discuss more)

isuffix commented on lib/tests/test_executable_bit.rs:
```diff
@@ -0,0 +1,158 @@
+fn prepare_exec_commits(ws: &TestWorkspace, repo_path: &RepoPath) -> (Commit, Commit) {
+    let tree = testutils::create_single_tree(&ws.repo, &[(repo_path, "")]);
+    let tree_no_exec = tree.id().clone();
+    let tree_exec = set_tree_executable(&tree, repo_path, true);
```
I'm replacing `set_tree_executable()` with a new `create_tree_executable()` function that effectively unrolls the various steps of `TestTreeBuilder` writing an empty file, and is called twice with true/false for the executable bits.

I'm not sure how we would get around writing trees twice since we still need two tree ids for the two commits? Let me know if there's something I missed!

isuffix commented on lib/src/local_working_copy.rs:
```diff
@@ -130,10 +130,10 @@ use crate::working_copy::WorkingCopyStateError;
 // ignore conditionally? #3949
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct FileExecutableFlag(#[cfg(unix)] bool);
+pub struct ExecBit(#[cfg(unix)] bool);
```
I'll update the commit description based on discussion in this thread since I assume you'll have thoughts.

Here's a draft of my reasons for the renaming:

1. `ExecBit` is fewer characters and far fewer syllables than `FileExecutableFlag`. This makes code more legible: both easier to mentally parse and quicker to type.

2. `File` in the name is unnecessary. `ExecBit` is already stored in the `FileType` inside the `FileState`, and the doc comment clarifies that it's the state for a file if you were unsure.

3. I don't care about the `Flag` vs. `Bit` part of the name. I'd be fine with `ExecFlag`/`exec_flag` instead if you prefer.

4. Most importantly, this solves the reuse of the `executable` name for variables that are both booleans and `FileExecutableFlag`. Using the same name for variables of different types gets confusing and caused me headaches keeping track of the source of values as they pass through the functions. Besides just renaming the type, this commit also adjusts variable names so we consistently name `ExecBit` variables `exec_bit` and boolean variables `executable`.

It would be worthwhile to make a commit just for the final point, but it gives us an excuse to rename the type as well to improve the first two points.

yuja replied:
The purpose of the `File` prefix was to clarify that `FileExecutableFlag` is NOT the executable flag stored in the backend `TreeValue`. We shouldn't compare `TreeValue`'s executable flag with `FileExecutableFlag`.

I don't care much about `Exec` vs `Executable` and `Flag` vs `Bit`.

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -14,6 +14,7 @@
 
 #![allow(missing_docs)]
 #![allow(clippy::let_unit_value)]
+#![allow(clippy::unnecessary_map_or)]
```
If `clippy::unnecessary_map_or` applies largely to that `impl` block, that's good. I don't know why we can't avoid `unnecessary_map_or`, though.

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -127,35 +127,104 @@ use crate::working_copy::WorkingCopy;
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct ExecBit(bool);
+pub struct ExecBit {
+    in_repo: bool,
+    #[cfg(unix)]
+    on_disk: bool,
```
> "Preserve the executable bit on all platforms" 

Can't we just copy the executable bit recorded before? The value wouldn't matter when the policy is `Ignore`. There would be no reliable way to update the recorded executable bit matching to the on-disk value if the backing filesystem doesn't support executable bit. (If the filesystem supports it, we could update the executable bit even if the policy is `Ignore`, though.)

yuja commented on lib/tests/test_local_working_copy_executable_bit.rs:
```diff
@@ -0,0 +1,159 @@
+            copy_id: CopyId::placeholder(),
+        },
+    );
+    tree_builder.write_tree().unwrap()
```
nit: use `TestTreeBuilder`. `tree_builder.file(repo_path, "").executable(executable); tree_builder.write_single_tree()`.

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -126,33 +126,65 @@ use crate::working_copy::WorkingCopy;
+        #[cfg(unix)]
+        return Self(in_repo);
+        #[cfg(windows)]
+        return Self(prev_on_disk().map_or(false, |prev| prev.0));
```
nit: `prev_on_disk().unwrap_or(Self(false))` instead of suppressing clippy.

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -1550,16 +1587,16 @@ impl FileSnapshotter<'_> {
-                    id: _,
                     executable,
+                    id: _,
                     copy_id: _,
```
nit: just leave this unchanged. The original code looks better as these fields are defined in that order.

isuffix replied:
Ope, this one was not intended.

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -1926,6 +1953,19 @@ impl TreeState {
+                // on-disk executable bit.
+                let get_prev_exec = || self.file_states().get_exec_bit(&path);
+                ExecBit::new_from_repo(executable, get_prev_exec)
+            };
```
This might be true on Unix if the policy is `Ignore`, but it wouldn't be on Windows. Because `write_file()` creates new file, the file wouldn't retain the old exec bit. Since `set_executable()` is noop on Windows, I think the created file should be considered non-executable.

FWIW, I think it's okay to set exec bit to disk even if the policy is `Ignore`.

yuja commented on lib/src/file_util.rs:
```diff
@@ -307,14 +310,59 @@ mod platform {
+            let supported = check_executable_bit_support(dir.path()).unwrap();
+            assert!(supported);
+        }
+    }
```
nit: delete this test as it is inherently unreliable? If we need some debugging function to test `check_executable_bit_support()`, I think we can add `jj debug` subcommand.

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -126,6 +126,37 @@ use crate::working_copy::WorkingCopy;
+        #[cfg(windows)]
+        return Self::Ignore;
+        #[cfg(unix)]
+        return match crate::file_util::check_executable_bit_support(wc_path) {
```
I don't think we should create random files in the working-copy directory. Perhaps, we can instead use the state directory? It might also be good to use named file and cache the detection result so readonly operation wouldn't attempt to change the on-disk content. For example, Mercurial caches the result in `.hg/wcache/checkisexec`.
https://repo.mercurial-scm.org/hg/file/7.1.1/mercurial/posix.py#l234

Since this PR is big, it's also good to split "auto" detection feature to separate PR.

isuffix replied:
I'll update to use the state path.

As for caching, I don't think it's necessary for the initial feature, and I'd rather not be in charge of adding it.

yuja replied:
> I'd rather not split this PR. The automatic checking is an integral part of this feature, and last time I split up a PR it just fragmented discussion and delayed things more.

Well, large PR tends to slow down review because the reviewer has to allocate time to concentrate on. Since this PR is useful even without auto detection of filesystem support, PR can be split up and will help improve review throughput.

yuja commented on lib/src/local_working_copy.rs:
```diff
@@ -126,6 +126,37 @@ use crate::working_copy::WorkingCopy;
+        return match crate::file_util::check_executable_bit_support(wc_path) {
+            Ok(false) => Self::Ignore,
+            Ok(true) => Self::Respect,
+            Err(_) => Self::Respect,
```
nit: propagate error or insert `tracing::warn!()` for debugging?

martinvonz commented:
I haven't looked at this yet but it would be nice to have better descriptions for commits 3-5. They don't say much about the motivation or whether there are observable differences.

yuja commented:
> ```shell
> jj config set --repo working-copy.exec-bit-change ignore
> touch file
> jj file chmod x file
> # on-disk file is not executable
> jj config set --repo working-copy.exec-bit-change respect
> jj file chmod x file
> # on-disk file should be executable
> ```

This is common problem in EOL handling and watchman, and I personally don't think we need a special support. User should check out all files if they change `exec-bit-change` mode.

If we need a workaround, we can record the previous `working-copy.exec-bit-change` mode in tree_state file. If the mode changes, cached file states cannot be trusted, so all working-copy files will be snapshotted. In the example above, the second `jj file chmod x file` will notice that the working-copy file isn't an executable, so it will first snapshot the current state and update it to executable.

martinvonz commented:
It sounds like this is ready to merge. Anything left to do, @isuffix?

isuffix commented:
Rebased on main and fixed the conflicts and broken tests.

> It sounds like this is ready to merge. Anything left to do, @isuffix?

As you can see from the error in CI, this is not ready. We still need to resolve the issue in my previous comment.

For that, we can either go back to storing both the `on-disk` and `in-repo` bits in the `FileState` or start storing the `working-copy.exec-bit-change` mode in the working copy's `tree_state` file. I prefer the former, but I'm fine with either. What would you prefer?

isuffix commented:
I'll merge this tomorrow unless there are any new comments.

isuffix commented:
Moved the changelog entry to unreleased (0.37)
